pkg/network/go/bininspect/dwarf.go

```diff
@@ -0,0 +1,325 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022-present Datadog, Inc.
+
+package bininspect
+
+import (
+	"debug/dwarf"
+	"debug/elf"
+	"errors"
+	"fmt"
+	"github.com/DataDog/datadog-agent/pkg/network/go/dwarfutils"
+	"github.com/DataDog/datadog-agent/pkg/network/go/dwarfutils/locexpr"
+	"github.com/go-delve/delve/pkg/dwarf/godwarf"
+	"github.com/go-delve/delve/pkg/dwarf/loclist"
+)
+
+// dwarfInspector is used to keep common data for the dwarf inspection functions.
+type dwarfInspector struct {
+	elfFile   *elf.File
+	arch      GoArch
+	dwarfData *dwarf.Data
+}
+
+// InspectWithDWARF returns the offsets of the given functions and fields in the given elf file.
+// It also returns some additional relevant metadata about the given file.
+// It is using the DWARF debug data to obtain information, and therefore should be run on elf files that contain debug
+// data, like our test binaries.
+func InspectWithDWARF(elfFile *elf.File, functions []string, structFields []FieldIdentifier) (*Result, error) {
+	if elfFile == nil {
+		return nil, ErrNilElf
+	}
+
+	// Determine the architecture of the binary
+	arch, err := GetArchitecture(elfFile)
+	if err != nil {
+		return nil, err
+	}
+
+	dwarfData, ok := HasDwarfInfo(elfFile)
+
+	if !ok || dwarfData == nil {
+		return nil, errors.New("expected dwarf data")
+	}
+
+	inspector := dwarfInspector{
+		elfFile:   elfFile,
+		arch:      arch,
+		dwarfData: dwarfData,
+	}
+
+	// Scan for functions and struct offsets
+	functionsMetadata, err := inspector.findFunctionsUsingDWARF(functions)
+	if err != nil {
```

Comment:
Should we also fail if we can't find all or some of the functions?

Reply:
yes I'll add a check